src/analysis/calculus/implicit.lean

```diff
@@ -0,0 +1,308 @@
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] [complete_space E]
+  {F : Type*} [normed_group F] [normed_space 𝕜 F] [complete_space F]
+  {G : Type*} [normed_group G] [normed_space 𝕜 G] [complete_space G]
+  (φ : implicit_function_data 𝕜 E F G)
+
+/-- The function given by `x ↦ (left_fun x, right_fun x)`. -/
+def prod_fun (x : E) : F × G := (φ.left_fun x, φ.right_fun x)
+
+@[simp] lemma prod_fun_apply (x : E) : φ.prod_fun x = (φ.left_fun x, φ.right_fun x) := rfl
+
+protected lemma has_strict_fderiv_at :
+  has_strict_fderiv_at φ.prod_fun
+    (φ.left_deriv.equiv_prod_of_surjective_of_is_compl φ.right_deriv φ.left_range φ.right_range
+       φ.is_compl_ker : E →L[𝕜] F × G) φ.pt :=
+φ.left_has_deriv.prod φ.right_has_deriv
+
+/-- Implicit function theorem. If `f : E → F` and `g : E → G` are two maps strictly differentiable
+at `a`, their derivatives `f'`, `g'` are surjective, and the kernels of these derivatives are
+complement subspaces of `E`, then `x ↦ (f x, g x)` defines a local homeomorphism between
+`E` and `F × G`. In particular, `f x = f a` is locally homeomorphic to `G`. -/
+def to_local_homeomorph : local_homeomorph E (F × G) :=
+φ.has_strict_fderiv_at.to_local_homeomorph _
+
+/-- Implicit function theorem. If `f : E → F` and `g : E → G` are two maps strictly differentiable
+at `a`, their derivatives `f'`, `g'` are surjective, and the kernels of these derivatives are
+complement subspaces of `E`, then `implicit_function_of_is_compl_ker` is the unique (germ of a) map
+`φ : F → G → E` such that `f (φ y z) = y` and `g (φ y z) = z`. -/
+def implicit_function : F → G → E := function.curry $ φ.to_local_homeomorph.symm
+
+@[simp] lemma to_local_homeomorph_coe : ⇑(φ.to_local_homeomorph) = φ.prod_fun := rfl
+
+lemma to_local_homeomorph_apply (x : E) :
+  φ.to_local_homeomorph x = (φ.left_fun x, φ.right_fun x) :=
+rfl
+
+lemma pt_mem_to_local_homeomorph_source :
+  φ.pt ∈ φ.to_local_homeomorph.source :=
+φ.has_strict_fderiv_at.mem_to_local_homeomorph_source
+
+lemma map_pt_mem_to_local_homeomorph_target :
+  (φ.left_fun φ.pt, φ.right_fun φ.pt) ∈ φ.to_local_homeomorph.target :=
+φ.to_local_homeomorph.map_source $ φ.pt_mem_to_local_homeomorph_source
+
+lemma prod_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.prod_fun (φ.implicit_function p.1 p.2) = p :=
+φ.has_strict_fderiv_at.eventually_right_inverse.mono $ λ ⟨z, y⟩ h, h
+
+lemma left_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.left_fun (φ.implicit_function p.1 p.2) = p.1 :=
+φ.prod_map_implicit_function.mono $ λ z, congr_arg prod.fst
+
+lemma right_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.right_fun (φ.implicit_function p.1 p.2) = p.2 :=
+φ.prod_map_implicit_function.mono $ λ z, congr_arg prod.snd
+
+lemma implicit_function_apply_image :
+  ∀ᶠ x in 𝓝 φ.pt, φ.implicit_function (φ.left_fun x) (φ.right_fun x) = x :=
+φ.has_strict_fderiv_at.eventually_left_inverse
+
+lemma to_implicit_function
```

Comment:
The name is strange, since you're not in the `has_strict_fderiv_at` namespace. I think `has_strict_fderiv_at` should be part of the name.

Reply:
```suggestion
lemma implicit_function_has_strict_fderiv_at
```